### PR TITLE
Feature/moscot not

### PR DESCRIPTION
**Pull request by MUCDK**

### CODE REVIEW

File: src/moscot/backends/ott/_jax_data.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+from ott.problems.linear import linear_problem
+from ott.solvers.linear import sinkhorn
+
+
+class JaxSampler:
+    """Data sampler for Jax."""
+
+    def __init__(
+        self,
+        distributions: List[jnp.ndarray],  # type: ignore[name-defined]
```
Comment (michalk8) [inline src/moscot/backends/ott/_jax_data.py:17]: I'd just use a list of `linear_problem.LinearProblem()`, this way, you can absorb the following arguments:
- distributions (contained in the `.geom` as a point cloud)
- a
- b
- tau_a
- tau_b
- epsilon (contained in the `.geom`)
- policy_pairs - can be constructed outside (or, e.g., have a `from_policy` class method).

File: src/moscot/backends/ott/_neuraldual.py [hunk 2]
```diff
@@ -0,0 +1,579 @@
...
+            if not self.pos_weights:
+                self.state_g = self.state_g.replace(params=self.clip_weights_icnn(self.state_g.params))
+            if iteration % self.log_freq == 0:
+                pretrain_logs["loss"].append(loss)
+        # load params of g into state_f
+        # this only works when f & g have the same architecture
+        self.state_f = self.state_f.replace(params=self.state_g.params)
+        return {"pretrain_logs": pretrain_logs}  # type:ignore[dict-item]
+
+    def train_neuraldual(
```
Comment (michalk8) [inline src/moscot/backends/ott/_neuraldual.py:289]: Consider splitting into smaller functions.

File: src/moscot/backends/ott/_utils.py [hunk 3]
```diff
@@ -38,9 +69,251 @@ def _compute_sinkhorn_divergence(
...
+
+    def update(self, val: float, n: int = 1) -> None:
+        """Update the meter."""
+        self.sum += val * n
+        self.count += n
+        self.avg = self.sum / self.count
+
+
+@partial(jax.jit, static_argnames=["k"])
+def get_nearest_neighbors(
```
Comment (michalk8) [inline src/moscot/backends/ott/_utils.py:92]: Consider using: https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.approx_min_k.html#jax.lax.approx_min_k

File: src/moscot/backends/ott/_utils.py [hunk 4]
```diff
@@ -38,9 +69,251 @@ def _compute_sinkhorn_divergence(
...
+        The second conditional dual potential function.
+    cost_fn
+        The cost function used to solve the OT problem.
+    corr
+        Whether the duals solve the problem in distance form, or correlation
+        form (as used for instance for ICNNs, see, e.g., top right of p.3 in
+        :cite:`makkuva:20`)
+    """
+
+    def __init__(self, state_f: TrainState, state_g: TrainState):
```
Comment (michalk8) [inline src/moscot/backends/ott/_utils.py:132]: Consider subclassing the `DualPotentials` directly and adding the condition here to init + use `f/g` local function as defined, e.g., in `to_dual_potentials`.

File: src/moscot/backends/ott/_utils.py [hunk 5]
```diff
@@ -38,9 +69,251 @@ def _compute_sinkhorn_divergence(
...
+
+def _get_optimizer(
+    learning_rate: float = 1e-3, b1: float = 0.5, b2: float = 0.9, weight_decay: float = 0.0, **kwargs: Any
+) -> Type[optax.GradientTransformation]:
+    return optax.adamw(learning_rate=learning_rate, b1=b1, b2=b2, weight_decay=weight_decay, **kwargs)
+
+
+# Compute the difference in drug signatures
+@jax.jit
+def compute_ds_diff(control, treated, push_fwd):
```
Comment (michalk8) [inline src/moscot/backends/ott/_utils.py:238]: Consider removing if not needed.

File: src/moscot/base/output.py [hunk 6]
```diff
@@ -369,3 +369,50 @@ def _ones(self, n: int) -> ArrayLike:
         import jax.numpy as jnp
 
         return jnp.ones((n,), dtype=self.transport_matrix.dtype)
+
+
+class BaseNeuralOutput(BaseSolverOutput, ABC):
```
Comment (michalk8) [inline src/moscot/base/output.py:374]: Consider merging `BaseNeuralOutput` and `BaseCondNeuralOutput` into one (making `condition` optional).

File: src/moscot/backends/ott/output.py [hunk 7]
```diff
@@ -178,30 +229,792 @@ def to(self, device: Optional[Device_t] = None) -> "OTTOutput":  # noqa: D102
...
+        Returns
+        -------
+        Backward potential evaluated at `x`.
+        """
+        if x.ndim not in (1, 2):
+            raise ValueError(f"Expected 1D or 2D array, found `{x.ndim}`.")
+        return jax.vmap(self._output.g)(cond, x)
+
+
+class GapNeuralOutput(ConvergencePlotterMixin, BaseNeuralOutput):
```
Comment (michalk8) [inline src/moscot/backends/ott/output.py:823]: I'd consider having a general `FunctionalOutput`, which needs an object that implements, e.g., `transport()` and `distance()` functions (this can be done via a class with abstract methods; in the case of NeuralDual, I'd subclass this class + ott's NeuralDual, which will then satisfy the interface.

File: src/moscot/backends/ott/solver.py [hunk 8]
```diff
@@ -307,3 +312,302 @@ def is_fused(self) -> Optional[bool]:
...
+    def _split_data(
+        self,
+        x: ArrayLike,
+        y: ArrayLike,
+        train_size: float,
+        seed: int,
+        a: Optional[ArrayLike] = None,
+        b: Optional[ArrayLike] = None,
+    ) -> Tuple[
+        ArrayLike,
```
Comment (michalk8) [inline src/moscot/backends/ott/solver.py:383]: Used named tuples.

File: src/moscot/utils/tagged_array.py [hunk 9]
```diff
@@ -159,3 +172,205 @@ def is_kernel(self) -> bool:
...
+            conditions_attr is not None and conditions_key is None
+        ):
+            raise ValueError(r"Either both `conditions_attr` and `conditions_key` must be `None` or none of them.")
+        return xy_attr is not None, xx_attr is not None, conditions_attr is not None
+
+    @classmethod
+    def from_adata(
+        cls,
+        adata: AnnData,
+        a: ArrayLike,
```
Comment (michalk8) [inline src/moscot/utils/tagged_array.py:258]: Why not use uniform marginals as defaults?

File: src/moscot/utils/tagged_array.py [hunk 10]
```diff
@@ -159,3 +172,205 @@ def is_kernel(self) -> bool:
...
+        backend
+            Backend to use.
+        kwargs
+            Keyword arguments to pass to the cost functions.
+
+        Returns
+        -------
+        The distribution container.
+        """
+        contains_linear, contains_quadratic, contains_condition = cls._verify_input(
```
Comment (michalk8) [inline src/moscot/utils/tagged_array.py:309]: Would consider adding these as properties, might be useful when using this container.

File: src/moscot/base/problems/problem.py [hunk 11]
```diff
@@ -808,3 +829,284 @@ def __repr__(self) -> str:
...
+    def solve(
+        self,
+        backend: Literal["ott"] = "ott",
+        device: Optional[Device_t] = None,
+        **kwargs: Any,
+    ) -> "OTProblem":
+        """Solve method."""
+        if self._xy is None:
+            raise ValueError("Unable to solve the problem without `xy`.")
+        return super().solve(backend=backend, device=device, cond_dim=0, input_dim=self._xy.data_src.shape[1], **kwargs)
```
Comment (michalk8) [inline src/moscot/base/problems/problem.py:847]: Any reason why using `self._xy` and not `self.xy`?

File: src/moscot/base/problems/problem.py [hunk 12]
```diff
@@ -808,3 +829,284 @@ def __repr__(self) -> str:
...
+        except KeyError:
+            raise KeyError(f"Unable to find data in `adata.obs[{policy_key!r}]`.") from None
+
+        self._policy = create_policy(policy, adata=self.adata, key=policy_key)
+        _ = self.policy.create_graph()  # type: ignore[union-attr]
+        self._sample_pairs = list(self.policy._graph)  # type: ignore[union-attr]
+
+        for el in self.policy.categories:  # type: ignore[union-attr]
+            mask = self._create_mask(el)
+            a_created = self._create_marginals(self.adata[mask], data=a, source=True, **kwargs)
```
Comment (michalk8) [inline src/moscot/base/problems/problem.py:1002]: Would create `adata_masked = self.adata[mask]` and use it multiple times.

File: src/moscot/utils/tagged_array.py [hunk 13]
```diff
@@ -159,3 +172,205 @@ def is_kernel(self) -> bool:
...
+            xx_data = None
+            xx_cost_fn = None
+
+        conditions_data = (
+            cls._extract_data(adata, attr=conditions_attr, key=conditions_key) if contains_condition else None  # type: ignore[arg-type]  # noqa:E501
+        )
+        return cls(xy=xy_data, xx=xx_data, a=a, b=b, conditions=conditions_data, cost_xy=xy_cost_fn, cost_xx=xx_cost_fn)
+
+
+class DistributionCollection:
```
Comment (michalk8) [inline src/moscot/utils/tagged_array.py:333]: I wonder how useful this class is, since it's just a like read-only dict with custom repr.
If you're planning to add more functionality to it (maybe like some sanity checks), would keep it.

File: src/moscot/backends/ott/solver.py [hunk 14]
```diff
@@ -340,3 +355,208 @@ def _call_kwargs(cls) -> Tuple[Set[str], Set[str]]:
...
+        **kwargs: Any,
+    ) -> Tuple[JaxSampler, JaxSampler]:
+        if xy.data_tgt is None:
+            raise ValueError(f"Unable to obtain target data from `xy={xy}`.")
+        x, y = self._assert2d(xy.data_src), self._assert2d(xy.data_tgt)
+        n, m = x.shape[1], y.shape[1]
+        if n != m:
+            raise ValueError(f"Expected `x/y` to have the same number of dimensions, found `{n}/{m}`.")
+        train_size = kwargs.pop("train_size", 1.0)
+        if train_size > 1.0 or train_size <= 0.0:
```
Comment (michalk8) [inline src/moscot/backends/ott/solver.py:384]: Would expose with defaults (also in other places).

File: src/moscot/backends/ott/solver.py [hunk 15]
```diff
@@ -340,3 +355,208 @@ def _call_kwargs(cls) -> Tuple[Set[str], Set[str]]:
...
+        train_a: List[Optional[ArrayLike]] = []
+        train_b: List[Optional[ArrayLike]] = []
+        valid_data: List[Optional[ArrayLike]] = []
+        valid_a: List[Optional[ArrayLike]] = []
+        valid_b: List[Optional[ArrayLike]] = []
+
+        sample_to_idx: Dict[Any, int] = {}
+        kwargs = _filter_kwargs(JaxSampler, **kwargs)
+        if train_size == 1.0:
+            train_data = [d.xy for d in distributions]  # type:ignore[var-annotated]
```
Comment (michalk8) [inline src/moscot/backends/ott/solver.py:495]: Think this is a bug, `d` is a key, not value in `distributions`.

File: src/moscot/backends/ott/solver.py [hunk 16]
```diff
@@ -340,3 +355,208 @@ def _call_kwargs(cls) -> Tuple[Set[str], Set[str]]:
...
+            train_a = [d.a for d in distributions]  # type:ignore[var-annotated]
+            train_b = [d.b for d in distributions]  # type:ignore[var-annotated]
+            valid_data, valid_a, valid_b = train_data, train_a, train_b
+            sample_to_idx = {k: i for i, k in enumerate(distributions)}
+        else:
+            if train_size > 1.0 or train_size <= 0.0:
+                raise ValueError("Invalid train_size. Must be: 0 < train_size <= 1")
+
+            seed = kwargs.pop("seed", 0)
+            for i, key in enumerate(distributions):
```
Comment (michalk8) [inline src/moscot/backends/ott/solver.py:505]: for i, (key, dist) in enumerate(distributions.items()):

File: src/moscot/backends/ott/solver.py [hunk 17]
```diff
@@ -340,3 +355,208 @@ def _call_kwargs(cls) -> Tuple[Set[str], Set[str]]:
...
+
+        self._train_sampler = JaxSampler(
+            train_data, sample_pairs, conditional=True, a=train_a, b=train_b, sample_to_idx=sample_to_idx, **kwargs
+        )
+        self._valid_sampler = JaxSampler(
+            valid_data, sample_pairs, conditional=True, a=valid_a, b=valid_b, sample_to_idx=sample_to_idx, **kwargs
+        )
+        return (self._train_sampler, self._valid_sampler)
+
+    def _split_data(  # type:ignore[override]
```
Comment (michalk8) [inline src/moscot/backends/ott/solver.py:529]: This bears a lot of similarity with the parent class's `_split_data`, would unify.

**Review by michalk8 — CHANGES_REQUESTED** [review]

File: src/moscot/utils/tagged_array.py (hunk 9 above)
Comment (MUCDK) [reply src/moscot/utils/tagged_array.py:258]: because we create them before already, so not necessary. Still, we could have a default value here, wdyt

File: src/moscot/backends/ott/_utils.py (hunk 4 above)
Comment (MUCDK) [reply src/moscot/backends/ott/_utils.py:139]: will adapt this once it's refactored in OTT-JAX

File: src/moscot/backends/ott/output.py [hunk 18]
```diff
@@ -178,30 +229,792 @@ def to(self, device: Optional[Device_t] = None) -> "OTTOutput":  # noqa: D102
...
+    @property
+    def potentials(  # type: ignore[override]
+        self,
+    ) -> Tuple[Callable[[jnp.ndarray], float], Callable[[jnp.ndarray], float]]:
+        """Return the learned potential functions."""
+        f = jax.vmap(self._output.f)
+        g = jax.vmap(self._output.g)
+        return f, g
+
+    def push(self, x: ArrayLike) -> ArrayLike:  # type: ignore[override]
```
[the remark this answers is not in the document]
Comment (MUCDK) [reply src/moscot/backends/ott/output.py:507]: I guess this is necessary as we have a check `x.shape[0] != self.shape[0]`. And it's hard to distinguish whether we should perform this check in `BaseSolverOutput.push/pull`

[the remark this answers is not in the document]
Comment (MUCDK) [reply src/moscot/backends/ott/output.py:507]: Alternatively, we could have a `FunctionalOutput` as you describe in the comment below. Should `FunctionalOutput` then inherit from `BaseSolverOutput`?

File: pyproject.toml [hunk 19]
```diff
@@ -58,12 +58,20 @@ dependencies = [
     "ott-jax>=0.4.4",
     "cloudpickle>=2.2.0",
     "rich>=13.5",
+    "optax>=0.1.7",
```
Comment (MUCDK) [inline pyproject.toml:61]: @michalk8 how do I make these optional, i.e. a `neural` optional dependency as below, but still make github CI install these optional dependencies?

**Comment by jannisborn** [comment]
Very nice effort! Is there any timeline on when this PR will be merged? Or, if you are looking for help, could you describe what is missing for this PR to be merged?

**Comment by hbenedek** [comment]
Hi @MUCDK!

I am working with @jannisborn and saw that you started to incorporate this PR to ott-jax (and GENOT and more). I would be more than happy to help out with testing/ giving feedback or contributing!

**Comment by MUCDK** [comment]
Hi @hbenedek ,
Great to hear you're interested in collaborating! Indeed, I would highly appreciate your help.
I think/hope we'll be done with the implementation on the ott-jax side by the end of this week. Still, please feel free to play around with the current implementation and provide feedback. This way, you'll get to know the API, and we can implement the models (i.e. OT-FM, GENOT, and eventually ICNN-OT (all in their conditional format, too) into moscot next week. Therefore, it'd be great if you are familiar with these models, as this will make the workflow easier!

Let's chat about details of the ott-jax -> moscot implementation in the beginning of next week!

**Comment by hbenedek** [comment]
Thanks @MUCDK! I will try to familiarize myself with the API and the models.
